### PR TITLE
raft: use leader's term when candidate becomes follower

### PR DESCRIPTION
**Pull request by july2993**
should fix https://github.com/pingcap/raft-rs/issues/34

### CODE REVIEW

File: src/raft.rs [hunk 1]
```diff
@@ -1503,22 +1503,27 @@ impl<T: Storage> Raft<T> {
...
-                info!("{} no leader at term {}; dropping proposal", self.tag, term);
+                info!(
+                    "{} no leader at term {}; dropping proposal",
+                    self.tag, self.term
+                );
                 return Err(Error::ProposalDropped);
             }
             MessageType::MsgAppend => {
-                self.become_follower(term, m.get_from());
+                debug_assert!(self.term == m.get_term());
```
Comment (BusyJay) [inline src/raft.rs]: Why not use `debug_assert_eq`?

File: tests/cases/test_raft.rs [hunk 2]
```diff
@@ -1971,6 +1971,69 @@ fn test_all_server_stepdown() {
...
+fn test_candidate_reset_term_msg_append() {
+    test_candidate_reset_term(MessageType::MsgAppend)
+}
+
+// test_candidate_reset_term tests when a candidate receives a
+// MsgHeartbeat or MsgAppend from leader, "step" resets the term
+// with leader's and reverts back to follower.
+fn test_candidate_reset_term(message_type: MessageType) {
+    let a = new_test_raft(1, vec![1, 2, 3], 10, 1, new_storage());
+    let b = new_test_raft(1, vec![1, 2, 3], 10, 1, new_storage());
```
Comment (BusyJay) [inline tests/cases/test_raft.rs]: Why 1?

File: tests/cases/test_raft.rs [hunk 3]
```diff
@@ -1971,6 +1971,69 @@ fn test_all_server_stepdown() {
...
+    assert_eq!(nt.peers[&2].state, StateRole::Follower);
+    assert_eq!(nt.peers[&3].state, StateRole::Follower);
+
+    // isolate 3 and increase term in rest
+    nt.isolate(3);
+    nt.send(vec![new_message(2, 2, MessageType::MsgHup, 0)]);
+    nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
+
+    // trigger campaign in isolated c
+    nt.peers
```
Comment (BusyJay) [inline tests/cases/test_raft.rs]: Should check state first.

File: src/raft.rs (hunk 1 above)
Comment (july2993) [reply src/raft.rs]: thanks @BusyJay updated all, debug_assert_eq is more suitable.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**